File: services/backend/python/src/api/tracing.py
"""Distributed tracing setup (ADR-0007): FastAPI + SQLAlchemy instrumentation,
exported via OTLP to the ADOT collector sidecar (which forwards to AWS X-Ray).
"""

from __future__ import annotations

from fastapi import FastAPI
from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
from opentelemetry.instrumentation.sqlalchemy import SQLAlchemyInstrumentor
from opentelemetry.sdk.resources import SERVICE_NAME, Resource
from opentelemetry.sdk.trace import SpanProcessor, TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor

from api.config import Settings
from api.db.engine import engine


def configure_tracing(
    app: FastAPI,
    settings: Settings,
    span_processor: SpanProcessor | None = None,
) -> None:
    """Instrument ``app`` and the DB engine for tracing, unless disabled.

    ``span_processor`` is an injection seam for tests (e.g. a
    ``SimpleSpanProcessor`` over an in-memory exporter); production code
    leaves it unset and gets the real OTLP-over-gRPC exporter.
    """
    if not settings.otel_traces_enabled:
        return

    provider = TracerProvider(resource=Resource.create({SERVICE_NAME: settings.otel_service_name}))
    processor = span_processor or BatchSpanProcessor(
        OTLPSpanExporter(endpoint=settings.otel_exporter_endpoint, insecure=True)
    )
    provider.add_span_processor(processor)
    trace.set_tracer_provider(provider)

    FastAPIInstrumentor.instrument_app(app)
    SQLAlchemyInstrumentor().instrument(engine=engine.sync_engine)

File: services/backend/go/internal/observability/tracing.go
package observability

import (
	"context"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	"go.opentelemetry.io/otel/trace"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"

	"github.com/iwata-jawsug-jp/devcon/services/backend/go/internal/config"
)

// Tracing bundles the tracer used to start spans and a flush hook the caller
// must invoke after each unit of work completes.
type Tracing struct {
	Tracer trace.Tracer
	// Flush exports any buffered spans immediately. Lambda may freeze the
	// execution environment between invocations, so — unlike a long-running
	// server that can rely on the batch processor's background export —
	// callers must invoke Flush at the end of every invocation or spans can be
	// lost. No-op when tracing is disabled.
	Flush func(context.Context) error
}

// ConfigureTracing sets up OTLP/gRPC tracing per ADR-0007 (SDK -> ADOT
// collector -> AWS X-Ray), unless cfg.OtelTracesEnabled is false (the
// default — mirrors services/backend/python/src/api/tracing.py). The ADOT
// endpoint itself (a Lambda extension in production) is wired up in Phase 3
// (#640); here the endpoint is just configuration.
func ConfigureTracing(ctx context.Context, cfg config.Settings) (Tracing, error) {
	if !cfg.OtelTracesEnabled {
		return Tracing{
			Tracer: otel.Tracer(cfg.OtelServiceName),
			Flush:  func(context.Context) error { return nil },
		}, nil
	}

	conn, err := grpc.NewClient(cfg.OtelExporterEndpoint, grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		return Tracing{}, err
	}

	exporter, err := otlptracegrpc.New(ctx, otlptracegrpc.WithGRPCConn(conn))
	if err != nil {
		return Tracing{}, err
	}

	res, err := resource.Merge(
		resource.Default(),
		resource.NewSchemaless(attribute.String("service.name", cfg.OtelServiceName)),
	)
	if err != nil {
		return Tracing{}, err
	}

	provider := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exporter),
		sdktrace.WithResource(res),
	)
	otel.SetTracerProvider(provider)

	return Tracing{
		Tracer: provider.Tracer(cfg.OtelServiceName),
		Flush:  provider.ForceFlush,
	}, nil
}
